src/target/riscv/riscv-011.c

```diff
@@ -455,12 +455,19 @@ static uint64_t dbus_read(struct target *target, uint16_t address)
 	dbus_status_t status;
 	uint16_t address_in;
 
+	// First scan sends the desired address/operation. This is 
+	// slightly inefficient as, if the previous read was to the same
+	// address, one could have simply used the data directly. However,
+	// that data may may have been captured significantly earlier,
+	// and could be considered stale at the start of this operation.
+	dbus_scan(target, &address_in, &value, DBUS_OP_READ, address, 0);
+
 	unsigned i = 0;
 	do {
 		status = dbus_scan(target, &address_in, &value, DBUS_OP_READ, address, 0);
```

Comment:
Also, I think if we go this route we need to change this to DBUS_OP_NOP. Otherwise, you won't be detecting the second BUSY that could be caused by *this* DBUS_OP_READ. This code (and the write code) assumes that the first operation always is accepted, which is only possible if it wasn't busy. In fact we should probably check that after the first dbus_scan.

Reply:
I just looked again at the dbus_write code. There you are explicitly checking that the write was accepted in one loop, then explicitly checking that the data is available in the second loop. I think you need the same construction in this dbus_read case (in theory, if dbus_read and dbus_write both use DBUS_OP_NOP for the second loop, there is no reason that you would ever *get* BUSY for the first loop, but otherwise it could be BUSY from the previous read).